### ISSUE TITLE
ParallelTests - Coverage Report Overwritten

### ISSUE THREAD

**Issue by sundarraja** [simplecov-ruby/simplecov#1019]
Test Frameworks: `Minitest` and `ParallelTests`
SimpleCov Version: `0.21.2`
Ruby Description: `ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux]`
Rails Version: `6.0.3.7`
Tests Invocation: `COVERAGE=true parallel_test -n 3 <space-separated-test-file-paths>`

Issue: In the screenshot below, the highlighted part should have been 'JSON Coverage Report generated for **(1/3), (2/3), and (3/3)**'. But the coverage result from the first 2 processes got overwritten somehow.

Issue Frequency: Not reproducible (Happens Randomly)

![coverage-report](https://user-images.githubusercontent.com/4964646/170007279-b16fab53-d5a7-4ded-8638-acf20b1bec3c.png)

**Comment by lluis** [comment]
it happens the same to me, with rspec + parallel_tests

run where report 12/12 overrides coverage:
```
lluis@laptop:~/git/vz (1685-test-coverage)$ rm -r coverage/ ; bundle exec rake parallel:spec
# (skipped non-coverage lines)
Coverage report generated for (3/12) to /home/lluis/git/vz/coverage. 5755 / 24443 LOC (23.54%) covered.
Coverage report generated for (1/12), (3/12) to /home/lluis/git/vz/coverage. 6570 / 23897 LOC (27.49%) covered.
Coverage report generated for (1/12), (3/12), (9/12) to /home/lluis/git/vz/coverage. 6861 / 23729 LOC (28.91%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (9/12) to /home/lluis/git/vz/coverage. 7244 / 23523 LOC (30.8%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 7963 / 23049 LOC (34.55%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (5/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 8551 / 22627 LOC (37.79%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (5/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 8941 / 22428 LOC (39.87%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (4/12), (5/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 9276 / 22230 LOC (41.73%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 9753 / 21953 LOC (44.43%) covered.
Coverage report generated for (1/12), (11/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 9920 / 21898 LOC (45.3%) covered.
Coverage report generated for (1/12), (10/12), (11/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 10042 / 21782 LOC (46.1%) covered.
Coverage report generated for (12/12) to /home/lluis/git/vz/coverage. 5295 / 24832 LOC (21.32%) covered.

```

correct run:

```
lluis@laptop:~/git/vz (1685-test-coverage)$ rm -r coverage ; bundle exec rake parallel:spec
Coverage report generated for (1/12) to /home/lluis/git/vz/coverage. 4994 / 25407 LOC (19.66%) covered.
Coverage report generated for (1/12), (3/12) to /home/lluis/git/vz/coverage. 5683 / 24936 LOC (22.79%) covered.
Coverage report generated for (1/12), (3/12), (9/12) to /home/lluis/git/vz/coverage. 6480 / 24503 LOC (26.45%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (9/12) to /home/lluis/git/vz/coverage. 6762 / 24331 LOC (27.79%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (6/12), (9/12) to /home/lluis/git/vz/coverage. 7274 / 24074 LOC (30.22%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (6/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 8642 / 22574 LOC (38.28%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (5/12), (6/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 9135 / 22266 LOC (41.03%) covered.
Coverage report generated for (1/12), (11/12), (3/12), (4/12), (5/12), (6/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 9329 / 22197 LOC (42.03%) covered.
Coverage report generated for (1/12), (11/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (9/12) to /home/lluis/git/vz/coverage. 9584 / 22086 LOC (43.39%) covered.
Coverage report generated for (1/12), (11/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 9798 / 21954 LOC (44.63%) covered.
Coverage report generated for (1/12), (10/12), (11/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 9961 / 21875 LOC (45.54%) covered.
Coverage report generated for (1/12), (10/12), (11/12), (12/12), (2/12), (3/12), (4/12), (5/12), (6/12), (7/12), (8/12), (9/12) to /home/lluis/git/vz/coverage. 10166 / 21722 LOC (46.8%) covered.

```

gem versions:
```
lluis@laptop:~/git/vz (1685-test-coverage)$ ruby -e "puts RUBY_DESCRIPTION"
ruby 2.6.2p47 (2019-03-13 revision 67232) [x86_64-linux]
lluis@laptop:~/git/vz (1685-test-coverage)$ bundle show simplecov
/home/lluis/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/simplecov-0.21.2
lluis@laptop:~/git/vz (1685-test-coverage)$ bundle show parallel_tests
/home/lluis/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/parallel_tests-3.11.0
lluis@laptop:~/git/vz (1685-test-coverage)$ bundle show rspec-rails
/home/lluis/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rspec-rails-3.8.3
lluis@laptop:~/git/vz (1685-test-coverage)$ bundle show capybara
/home/lluis/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/capybara-3.35.3
lluis@laptop:~/git/vz (1685-test-coverage)$ bundle show rails
/home/lluis/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rails-4.2.11.3
```

**Comment by Canertasan** [comment]
Same issue here ➕

**Comment by tjallingvanderwal** [comment]
### Observation

Also using `parallel_test`. 
Running on CircleCI.
Using `SimpleCov::Formatter::CoberturaFormatter` to generate `coverage.xml` before uploading results to [codecov.io](codecov.io).

Caught the following outout at the end of a test job:

```
Coverage report generated for (1/8), (2/8), (3/8), (4/8), (5/8), (6/8), (7/8), (8/8) to /home/circleci/********/coverage/coverage.xml. 19217 / 25717 LOC (74.72%) covered.
Coverage report generated for (1/8), (2/8), (3/8), (4/8), (5/8), (6/8), (8/8) to /home/circleci/********/coverage/coverage.xml. 18241 / 25690 LOC (71.0%) covered.
```

This is the same as what others have reported here.

### Hypothesis

Looking through the code, my hypothesis is that `.resultset.json` is protected by a file lock when reading/merging results. But anything that happens after that (anything any `Formatter` does), is unprotected. 

```
process A calls `store_result`                          # `.resultset.json` protected by file lock
process A calls `merged_result`                         # `.resultset.json` protected by file lock
process A holds @results in memory
process B calls `store_result`                          # `.resultset.json` protected by file lock
process B calls `merged_result`                         # `.resultset.json` protected by file lock
process B holds @results in memory
process B calls `at_exit { SimpleCov.result.format! }`  # `coverage.xml` not protected
process A calls `at_exit { SimpleCov.result.format! }`  # `coverage.xml` not protected
```

ref: https://github.com/simplecov-ruby/simplecov/blob/efdb08db63b35577b3b0d79db893ead1d848d8dd/lib/simplecov.rb#L107-L115



### Workaround / Our Solution

Let every parallel job write to a different directory:

```ruby
if ENV['CI']
  SimpleCov.coverage_dir "coverage/#{ENV['CI_NODE_INDEX']}/"
end
```

If the tooling (or cloud service) you use to track coverage can handle multiple coverage reports, then this is already enough.

Otherwise, use a custom script aggregate everything using [`SimpleCov.collate`](https://github.com/simplecov-ruby/simplecov/tree/efdb08db63b35577b3b0d79db893ead1d848d8dd#merging-test-runs-under-different-execution-environments).

**Comment by ahukkanen** [comment]
I am not very familiar with the inner workings of Simplecov but we are also experiencing this issue at [decidim/decidim](http://github.com/decidim/decidim) and based on this conversation I found the file locking at the `ResultMerge` suspicious:
https://github.com/simplecov-ruby/simplecov/blob/efdb08db63b35577b3b0d79db893ead1d848d8dd/lib/simplecov/result_merger.rb#L153-L166

It sure locks the file for the running process but it is not checking the file lock for any other processes which to me seems that if two processes are writing their results at the same time, it would get mixed.

As said, I am no expert with the Simplecov codebase but this just looked suspicious to me so I thought I'd mention it here.

Our run environment is basically the same as already reported here:
- parallel_tests
- simplecov
- simplecov-cobertura (as the CI formatter)

**Comment by knovoselic** [comment]
As far as I can see, there are no bugs related to the locking logic. I've had the same issue, and the problem turned out to be in the `SimpleCov.merge_timeout` option. By default, its set to 10 minutes, meaning that any runs/code coverage results which are older than 10 minutes will not get merged into the final report. Maybe it helps someone :)

**Comment by rangerscience** [comment]
I think I've run into this - we're using `parallel_test`, and when I run `bundle exec rake parallel:spec` I reliably get what seem to be unmerged coverage reports.

I dug into SimpleCov just a wee bit (in the ReadMe, it shows you the commit that's supposed to support parallel test), and [this bit](https://github.com/jshraibman-mdsol/simplecov/blob/10186bca60c3a2b2690f15856d52e24bc4c0c1f7/lib/simplecov/command_guesser.rb#L21) jumped out at me:
```
if ENV['PARALLEL_TEST_GROUPS'] && ENV['TEST_ENV_NUMBER']
```

I poked around inside of `parallel_test`, and it _does_ set that env var at some point (it also unsets it), so I suspect that this is a gem ordering issue. Regardless, running the tests with the following seems to have resolved the issue:

```
PARALLEL_TEST_GROUPS='' bundle exec rake parallel:spec
```

**Comment by huda-kh** [comment]
Having the same issue, no fix yet?

**Comment by kylesnowschwartz** [comment]
I ran into a similar issue where SimpleCov results were being overwritten when running tests in parallel across separate Buildkite processes. To solve this, I ensured each parallel test process had a unique command name and properly merged the coverage results at the end of all parallel runs.

Since this seems to be a common challenge in CI environments with multiple parallel jobs, I wanted to share a more generic version of my solution. This script ensures that SimpleCov properly collects and merges results when running in a parallelized test environment, whether it’s Buildkite, GitHub Actions, or another CI system.

Key features:
	•	Uses TEST_ENV_NUMBER to distinguish parallel processes.
	•	Ensures each process has a unique command name.
	•	Merges results at the end of all parallel processes.
	•	Uses ParallelTests.last_process? to determine when to trigger collation.

```
# frozen_string_literal: true

require 'simplecov'
require 'simplecov-console'
require 'fileutils'
require 'parallel_tests' # Ensure you have this gem if running parallel tests

# Set minimum coverage threshold from environment variable (optional)
MIN_COVERAGE = ENV.fetch('MIN_COVERAGE', nil)&.to_i

if MIN_COVERAGE
  SimpleCov.start do
    puts "Starting SimpleCov with a minimum coverage of #{MIN_COVERAGE}%"

    # Ensure a unique command name for each parallel test process
    command_name "TestSuite#{ENV.fetch('TEST_ENV_NUMBER', '')}"
    puts "Running Test Suite: #{command_name}"

    enable_coverage :branch

    # Customize these filters based on your project structure
    add_filter '/spec/'
    add_filter '/config/'
    add_filter '/vendor/' # Ignore external dependencies if applicable

    # Define coverage groups (customize as needed)
    add_group 'Models', 'app/models'
    add_group 'Controllers', 'app/controllers'
    add_group 'Libraries', 'lib'
  end

  # Enforce minimum coverage levels
  SimpleCov.minimum_coverage line: MIN_COVERAGE, branch: MIN_COVERAGE

  # Handle merging coverage results in parallel test environments
  if ENV['TEST_ENV_NUMBER'] && ParallelTests.last_process?
    ParallelTests.wait_for_other_processes_to_finish
    SimpleCov.at_exit do
      SimpleCov.result # Ensure last parallel process saves results
      SimpleCov.at_exit do
        puts 'Merging parallel coverage reports...'
        SimpleCov.result.format!
      end
      SimpleCov.collate Dir['coverage*/.resultset.json'] do
        SimpleCov.formatter SimpleCov::Formatter::MultiFormatter.new(
          [SimpleCov::Formatter::HTMLFormatter, SimpleCov::Formatter::Console]
        )
        SimpleCov::Formatter::Console.output_style = 'block'
      end
    end
  elsif ENV['TEST_ENV_NUMBER'] # Midway parallel process
    SimpleCov.at_exit { SimpleCov.result }
  else # Standard non-parallel execution
    SimpleCov.print_error_status = false
    SimpleCov.formatter SimpleCov::Formatter::SimpleFormatter
  end
end
```

**Comment by EduardoRSeifert** [comment]
I had this issue, but in my case it was a problem with the merge timeout. more specifically, i had a test using a gem to change the default time, and the test didn't return the time to the correct value after, resulting in one of the threads having "Time.now" set to tomorrow, which would say that the other coverage reports were 1 day younger than it and, therefore, not merge with them. to check if that's the case for you guys as well, you can monkey patch [this](https://github.com/simplecov-ruby/simplecov/blob/ce45fa65808a4811074f818f15df0d83d04b5185/lib/simplecov/result_merger.rb#L74) method to check the timestamp of your results using some print statements.

I'm using rails with rspec and parallel_tests, using the SimpleCov.start 'rails' config with an increased timeout only
